### PR TITLE
Implement a new Variant type for best-of-n sampling

### LINKED ISSUE #287: Add Rejection Sampling variant type

### PR DESCRIPTION
**Pull request by virajmehta**
There could still be many more integration tests written but I think this is in fairly good shape and worth looking at. Happy to write more / make changes as needed. When we merge, this will close #287

### CODE REVIEW

File: gateway/src/variant/rejection_sampling.rs [hunk 1]
```diff
@@ -0,0 +1,989 @@
...
+        }
+
+        Ok(selected_candidate)
+    }
+}
+
+/// Actually does the work of selecting the best candidate for rejection sampling.
+/// We factor this into a separate function so that we can gracefully handle any error here in the caller.
+/// If a model inference actually occurs, we return None and the model inference
+async fn inner_select_best_candidate<'a, 'request>(
```
Comment (GabrielBianconi) [inline gateway/src/variant/rejection_sampling.rs:299]: Document implementation and return type better.

File: gateway/src/variant/rejection_sampling.rs [hunk 2]
```diff
@@ -0,0 +1,989 @@
...
+                candidates.len()
+            ),
+        };
+        err.log();
+        return Ok((None, model_inference_result));
+    }
+    Ok((Some(answer_choice as usize), model_inference_result))
+}
+
+impl EvaluatorConfig {
```
Comment (GabrielBianconi) [inline gateway/src/variant/rejection_sampling.rs:358]: Document functions

File: gateway/src/variant/mod.rs [hunk 3]
```diff
@@ -12,14 +12,15 @@ use crate::minijinja_util::TemplateConfig;
...
-// pub mod rejection_sampling;
+pub mod rejection_sampling;
 
 #[derive(Debug, Deserialize)]
 #[serde(tag = "type")]
 #[serde(rename_all = "snake_case")]
 #[serde(deny_unknown_fields)]
 pub enum VariantConfig {
     ChatCompletion(chat_completion::ChatCompletionConfig),
+    RejectionSampling(rejection_sampling::RejectionSamplingConfig),
```
Comment (GabrielBianconi) [inline gateway/src/variant/mod.rs:23]: serde rename to `experimental_rejection_sampling`

**Comment by virajmehta** [comment]
TODO: change the name to best-of-n sampling

File: gateway/src/variant/best_of_n.rs [hunk 4]
```diff
@@ -293,9 +293,18 @@ impl RejectionSamplingConfig {
     }
 }
 
-/// Actually does the work of selecting the best candidate for rejection sampling.
-/// We factor this into a separate function so that we can gracefully handle any error here in the caller.
-/// If a model inference actually occurs, we return None and the model inference
+/// Attempts to select the best candidate for best of n.
+/// If this function returns an error or the index is None, we will randomly select one
+/// of the candidates in the outer function.
+/// If a model inference actually occurs, we return None and the model inference result instead of Err() so
```
Comment (GabrielBianconi) [inline gateway/src/variant/best_of_n.rs:299]: `If an issue occurs but a model inference...`

**Review by GabrielBianconi — APPROVED** [review]
Almost there. Let's merge pending empirical validation.

File: gateway/src/variant/best_of_n.rs [hunk 5]
```diff
@@ -0,0 +1,1014 @@
...
+    if answer_choice >= candidates.len() {
+        let err = Error::Inference {
+            message: format!(
+                "The index chosen by the evaluator is out of bounds: {} >= {}",
+                answer_choice,
+                candidates.len()
+            ),
+        };
+        err.log();
+        return Ok((None, model_inference_result));
```
Comment (virajmehta) [inline gateway/src/variant/best_of_n.rs:362]: log all failures here

File: examples/gsm8k-custom-recipe-dspy/gsm8k_dspy.ipynb [hunk 6]
```diff
@@ -88,15 +88,16 @@
...
     "    variant_name: Optional[str] = None,\n",
     "    dryrun: bool = False,\n",
-    ") -> bool:\n",
+    ") -> Optional[bool]:\n",
     "    response = await solve_math_problem(\n",
     "        example[\"question\"], client, variant_name=variant_name, dryrun=dryrun\n",
     "    )\n",
     "    if response is None:\n",
-    "        return False\n",
+    "        print(\"response is None\")\n",
```
Comment (GabrielBianconi) [inline examples/gsm8k-custom-recipe-dspy/gsm8k_dspy.ipynb:96]: .

File: examples/gsm8k-custom-recipe-dspy/gsm8k_dspy.ipynb [hunk 7]
```diff
@@ -168,7 +169,11 @@
     "    run_inference(example, variant_name=variant_name, dryrun=True)\n",
     "    for example in test_examples[:NUM_TEST_INFERENCES]\n",
     "]\n",
+    "\n",
     "results = await tqdm_asyncio.gather(*coroutines, desc=\"Running test inferences\")\n",
+    "# Filter out None values from results\n",
+    "results = [result for result in results if result is not None]\n",
+    "\n",
```
Comment (GabrielBianconi) [inline examples/gsm8k-custom-recipe-dspy/gsm8k_dspy.ipynb:176]: should report success rate

File: gateway/src/variant/best_of_n.rs [hunk 8]
```diff
@@ -311,9 +313,9 @@ async fn inner_select_best_candidate<'a, 'request>(
     models: &'a HashMap<String, ModelConfig>,
     inference_config: &'request InferenceConfig<'request>,
     client: &'request Client,
-    candidates: &Vec<InferenceResult<'request>>,
+    candidates: &[InferenceResult<'request>],
 ) -> Result<(Option<usize>, ModelInferenceResponseWithMetadata<'a>), Error> {
-    let inference_request = evaluator.prepare_request(
+    let (inference_request, skipped_indices) = evaluator.prepare_request(
```
Comment (GabrielBianconi) [inline gateway/src/variant/best_of_n.rs:318]: If there is a single candidate, we can skip the evaluator.

File: gateway/tests/e2e/best_of_n.rs [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+/// We check that the good response is selected and that the other responses are not
+/// but they get stored to the ModelInference table.
+#[tokio::test]
+async fn e2e_test_best_of_n_dummy_candidates_real_judge() {
+    let episode_id = Uuid::now_v7();
+
+    let payload = json!({
+        "function_name": "best_of_n",
+        "episode_id": episode_id,
+        "input":{
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/best_of_n.rs:23]: space

File: gateway/tests/e2e/best_of_n.rs [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+
+    let payload = json!({
+        "function_name": "best_of_n_json",
+        "episode_id": episode_id,
+        "input":{
+            "system": {"assistant_name": "AskJeeves"},
+            "messages": [
+                {
+                    "role": "user",
+                    "content": "What's the first word in the typical output of ones first program. Answer as a json object with a single field 'answer' containing the string."
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/best_of_n.rs:215]: one's

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]
almost there

**Review by GabrielBianconi — APPROVED** [review]

**Pull request merged**